### PR TITLE
Random Beacon v2 DKG challenge verification and slashing

### PR DESCRIPTION
**Pull request by lukasz-zimnoch**
Depends on: https://github.com/keep-network/sortition-pools/pull/133

This pull request introduces actual DKG challenge verification. The verification code calls the sortition pool to select the group using the DKG execution seed and computes a hash from the group members list. Then, it compares that hash with the hash computed from the submitted group members list. In case those hashes differ, the DKG challenge is deemed justified. A justified challenge causes the slashing of the members who signed the malicious result and triggers a notifier reward payment.

### CODE REVIEW

File: solidity/random-beacon/contracts/RandomBeacon.sol
Comment (lukasz-zimnoch) [inline solidity/random-beacon/contracts/RandomBeacon.sol:502]: @pdyraga Can you confirm this slashing and rewarding logic is correct? The RFC says " the notifier receives 5% from the slashed amount". If I understand correctly, slashing will burn 100% of tokens so the `RandomBeacon` must pay the reward from its balance. The reward amount is 5% of the total (?) slashing amount. Also, we can't transfer directly because there is no guarantee we have enough balance so making just an allowance seems the right option. We can also make the transfer only if the balance is good and the allowance otherwise.

Comment (pdyraga) [reply solidity/random-beacon/contracts/RandomBeacon.sol:502]: Nope, this is not correct. We should use `seize` function of the staking contract, see [here](https://github.com/threshold-network/solidity-contracts/blob/fa83b98566b7c786a499fc438d97b4dd3eee0293/contracts/staking/TokenStaking.sol#L893-L900). It will seize tokens and pay reward to `notifier` address. We do not need to care about `RandomBeacon` balance because the balance of this contract is just the maintenance pool and is not used for the reward of relay entry timeout or malicious DKG result notifier.

Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/RandomBeacon.sol:502]: Yep, I wasn't aware of `seize`. Will fix.

Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/RandomBeacon.sol]: And fixed: https://github.com/keep-network/keep-core/pull/2695/commits/e0e765b52541dc79d61274f6ca8f09675096bfbc

**Comment by lukasz-zimnoch** [comment]
@pdyraga Ready for another look!

**Comment by pdyraga** [comment]
@lukasz-zimnoch Can you solve merge conflicts? Would love to have the code in the final form before the review.

**Comment by lukasz-zimnoch** [comment]
> @lukasz-zimnoch Can you solve merge conflicts? Would love to have the code in the final form before the review.

Done @pdyraga

File: solidity/random-beacon/contracts/RandomBeacon.sol
Comment (pdyraga) [inline solidity/random-beacon/contracts/RandomBeacon.sol:217]: I'd rather say `DkgMaliciousResultSlashed` because it's the result that was malicious, not the slashing 😉

File: solidity/random-beacon/contracts/RandomBeacon.sol
Comment (pdyraga) [inline solidity/random-beacon/contracts/RandomBeacon.sol:594]: This 5% should be governable, just like `relayEntryTimeoutNotificationRewardMultiplier` is governable. 5% in the RFC was suggested before we knew how T staking contract should look like. T staking contract does not give that 5% from `slashingAmount` just like KEEP staking contract was doing but instead, it is giving it from a special notifier reward pool.

File: solidity/random-beacon/contracts/RandomBeacon.sol
Comment (pdyraga) [inline solidity/random-beacon/contracts/RandomBeacon.sol:580]: `resultHash` is only used in the event, do we need to cache it this way? Shouldn't we inline the call where we emit `DkgResultMaliciousSlashed`?

Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/RandomBeacon.sol:580]: Yep, we need it. Some cleanup is done within `dkg.challengeResult()`.

File: solidity/random-beacon/contracts/RandomBeacon.sol
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/RandomBeacon.sol:217]: See https://github.com/keep-network/keep-core/pull/2695/commits/f0db676b0854612c89aabe9532493efb07dbd660

File: solidity/random-beacon/contracts/RandomBeacon.sol
Comment (pdyraga) [reply solidity/random-beacon/contracts/RandomBeacon.sol:580]: Can we have challengeResult returning the hash? I think this is cleaner - we do not have to look inside about what information DKG library stores and when it is cleared up.

File: solidity/random-beacon/contracts/libraries/DKG.sol
Comment (pdyraga) [inline solidity/random-beacon/contracts/libraries/DKG.sol:351]: We do not need to fetch the addresses of all members, addresses of signing group members should be enough. Do we know what is the gas cost difference when working with a real sortition pool? Fetching just signing member addresses will require at least one iteration through `signingMembersIndices` and building an array with their identifiers but maybe it's still cheaper than doing more SREADs on the sortition pool side.

File: solidity/random-beacon/contracts/libraries/DKG.sol
Comment (pdyraga) [inline solidity/random-beacon/contracts/libraries/DKG.sol:212]: @nkuba Was there any particular reason it was done this way? Gas costs maybe?

File: solidity/random-beacon/test/RandomBeacon.GroupCreation.test.ts
Comment (pdyraga) [inline solidity/random-beacon/test/RandomBeacon.GroupCreation.test.ts:275]: s/an other actual group/another group ?

I think this will be easier to reason about

File: solidity/random-beacon/contracts/RandomBeacon.sol
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/RandomBeacon.sol]: Sure: https://github.com/keep-network/keep-core/pull/2695/commits/7c9f436788770bf52471df6721aa35ce6aa6f937

File: solidity/random-beacon/contracts/libraries/DKG.sol
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/libraries/DKG.sol:351]: We don't have a real sortition pool plugged-in yet to make a fair comparison. Maybe we can leave a TODO and check this while doing the optimization work?

File: solidity/random-beacon/contracts/libraries/DKG.sol
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/libraries/DKG.sol:212]: Can't compare gas usage as the original version errors out with `Stack too deep` at compilation stage 🤷

Comment (pdyraga) [reply solidity/random-beacon/contracts/libraries/DKG.sol:212]: Problem solved then.

File: solidity/random-beacon/test/RandomBeacon.GroupCreation.test.ts
Comment (pdyraga) [inline solidity/random-beacon/test/RandomBeacon.GroupCreation.test.ts:1685]: We need to lower or upper-case all revert messages one day. Not something for this PR, just complaining as always.

File: solidity/random-beacon/contracts/libraries/DKG.sol
Comment (pdyraga) [reply solidity/random-beacon/contracts/libraries/DKG.sol:351]: Fine for me, let's leave a TODO.

**Comment by pdyraga** [comment]
What is the gas cost of `submitDkgResult`, `approveDkgResult`, and `challengeDkgResult` compared to their versions in `main`?

File: solidity/random-beacon/contracts/RandomBeacon.sol
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/RandomBeacon.sol]: See https://github.com/keep-network/keep-core/pull/2695/commits/2f4e1b12c7674db2500142e70b384527b5f4afd7

File: solidity/random-beacon/contracts/libraries/DKG.sol
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/libraries/DKG.sol:355]: See https://github.com/keep-network/keep-core/pull/2695/commits/52a1623ca0747be3944ad5e3fe060e3d9b6b05cf

File: solidity/random-beacon/contracts/RandomBeacon.sol
Comment (pdyraga) [inline solidity/random-beacon/contracts/RandomBeacon.sol:155]: Actually, this should say "Percentage of the staking contract malicious behavior notification reward...", same for `relayEntryTimeoutNotificationRewardMultiplier`. Can we fix it? 🙏🏻

Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/RandomBeacon.sol]: Sure: https://github.com/keep-network/keep-core/pull/2695/commits/0d6f0ab25225ac5f2407f3a0d07b2b922fd33c89

File: solidity/random-beacon/test/RandomBeacon.GroupCreation.test.ts
Comment (lukasz-zimnoch) [reply solidity/random-beacon/test/RandomBeacon.GroupCreation.test.ts]: See https://github.com/keep-network/keep-core/pull/2695/commits/0d6f0ab25225ac5f2407f3a0d07b2b922fd33c89

**Comment by lukasz-zimnoch** [comment]
> What is the gas cost of `submitDkgResult`, `approveDkgResult`, and `challengeDkgResult` compared to their versions in `main`?



> What is the gas cost of `submitDkgResult`, `approveDkgResult`, and `challengeDkgResult` compared to their versions in `main`?

| Method               | Before (average) | After (average) |
|----------------------|------------------|-----------------|
| `submitDkgResult`    | 1182702          | 1727923         |
| `approveDkgResult`   | 144961           | 164681          |
| `challengeDkgResult` | 79012            | 512847          |

Bear in mind `challengeDkgResult` is just a starting point in `main`.

**Comment by pdyraga** [comment]
I am fine with `challengeDkgResult` being more expensive but the cost difference for `submitDkgResult` is too high. Assuming 115 Gwei gas price today - and it's not the highest gas price nowadays - plus $4237 ETH, it was $576 to submit DKG result and now it is $841 😬 I think 1.1M was already too high.

What makes it so expensive? Storing seed, the hash of submitted members, or computing that hash? Instead of storing the hash, we can read information from `Groups` library in `RandomBeacon.challengeDkgResult` and pass it to `DKG` library. This way, non-happy-path challenger pays and the happy path is less expensive. Though I don't know if this is the problematic piece.

**Comment by lukasz-zimnoch** [comment]
> I am fine with `challengeDkgResult` being more expensive but the cost difference for `submitDkgResult` is too high. Assuming 115 Gwei gas price today - and it's not the highest gas price nowadays - plus $4237 ETH, it was $576 to submit DKG result and now it is $841 😬 I think 1.1M was already too high.
> 
> What makes it so expensive? Storing seed, the hash of submitted members, or computing that hash? Instead of storing the hash, we can read information from `Groups` library in `RandomBeacon.challengeDkgResult` and pass it to `DKG` library. This way, non-happy-path challenger pays and the happy path is less expensive. Though I don't know if this is the problematic piece.

We can't get group members from `Groups` library. This is because members are set through `setGroupMembers` which changes the group members order in case there is at least one misbehaved member. But this is not the problematic part. A bigger problem is storing IDs of members who signed the result. We need to store it and use it in case a justified challenge emerges because we need to know who signed the malicious result and should be slashed. Slashing all the group will be cheaper but not fair against honest members. Anyway, `submitDkgResult` is expensive even without changes from this PR and it is still growing (min stake check in progress) so simple optimizations may not be enough. I'm afraid a bigger refactoring will be required.

**Comment by pdyraga** [comment]
> > I am fine with `challengeDkgResult` being more expensive but the cost difference for `submitDkgResult` is too high. Assuming 115 Gwei gas price today - and it's not the highest gas price nowadays - plus $4237 ETH, it was $576 to submit DKG result and now it is $841 😬 I think 1.1M was already too high.
> > What makes it so expensive? Storing seed, the hash of submitted members, or computing that hash? Instead of storing the hash, we can read information from `Groups` library in `RandomBeacon.challengeDkgResult` and pass it to `DKG` library. This way, non-happy-path challenger pays and the happy path is less expensive. Though I don't know if this is the problematic piece.
> 
> We can't get group members from `Groups` library. This is because members are set through `setGroupMembers` which changes the group members order in case there is at least one misbehaved member. But this is not the problematic part. A bigger problem is storing IDs of members who signed the result (64 `uint32` in the worst case). We need to store it and use it in case a justified challenge emerges because we need to know who signed the malicious result and should be slashed. Slashing all the group will be cheaper but not fair against honest members. Anyway, `submitDkgResult` is expensive even without changes from this PR and it is still growing (min stake check in progress) so simple optimizations may not be enough. I'm afraid a bigger refactoring will be required.

Let's then not store anything and expect the challenger/approver to pass the required data as parameters. We have `submittedResultHash` stored and it's a hash of the entire result. We do not need to store `submittedResultSigningMembers` and`submittedResultMisbehavedMembers`. If needed, the challenger can even pass the entire `Dkg.Result`. We just need to ensure events emitted during `submitDkgResult` have all the information required so that the challenger can assemble parameters. Then, in `challengeDkgResult`, we hash everything, and if the hash matches, use values provided by the challenger.

**Review by pdyraga — APPROVED** [review]
`main` now:
```
···························|·······························································|·············|·············|·············|···············|··············
|  RandomBeaconStub        ·  approveDkgResult                                             ·     121950  ·     171117  ·     131209  ·           72  ·          -  │
···························|·······························································|·············|·············|·············|···············|··············
|  RandomBeaconStub        ·  challengeDkgResult                                           ·      68986  ·      82666  ·      81653  ·           27  ·          -  │
···························|·······························································|·············|·············|·············|···············|··············
|  RandomBeaconStub        ·  submitDkgResult                                              ·     593915  ·     947537  ·     845238  ·          126  ·          -  │
···························|·······························································|·············|·············|·············|···············|··············
```

with this change:
```
···························|································································|·············|·············|·············|···············|··············
|  RandomBeaconStub        ·  approveDkgResult                                              ·     203923  ·     252442  ·     213151  ·           72  ·          -  │
···························|································································|·············|·············|·············|···············|··············
|  RandomBeaconStub        ·  challengeDkgResult                                            ·     987440  ·    1004540  ·     992580  ·           35  ·          -  │
···························|································································|·············|·············|·············|···············|··············
|  RandomBeaconStub        ·  submitDkgResult                                               ·     731782  ·    1057792  ·     996681  ·          130  ·          -  │
···························|································································|·············|·············|·············|···············|··············
```

151443 difference for `submitDkgResult` is a lot. I do not know what is the reason but I think computing the hash of `Result`, especially `signatures` could cost a lot. I would consider extracting `signatures` out of the `Result` and providing them to `submitDkgResult` separately given that I think they are not needed in any other function, are already validated, and do not have to be included in the result hash.

Let's also update this doc:
```
// `\x19Ethereum signed message:\n${keccak256(groupPubKey,misbehaved,startBlock)}`
bytes signatures;
```
as it is no longer up to date.

I think this change is good and needs some optimization love later. Merging to unblock everything else.

File: solidity/random-beacon/contracts/libraries/DKG.sol
Comment (pdyraga) [inline solidity/random-beacon/contracts/libraries/DKG.sol:119]: Let's also add a note that the final order of members in the group might be different after DKG completes. See `Groups.setGroupMembers`. Clients should `getGroup` after DKG to see what is their final group member ID in that group.

File: solidity/random-beacon/.solhintignore
Comment (pdyraga) [reply solidity/random-beacon/.solhintignore:2]: Done in https://github.com/keep-network/keep-core/pull/2734

File: solidity/random-beacon/contracts/libraries/DKG.sol
Comment (pdyraga) [reply solidity/random-beacon/contracts/libraries/DKG.sol:119]: Done in https://github.com/keep-network/keep-core/pull/2734

File: solidity/random-beacon/test/RandomBeacon.GroupCreation.test.ts
Comment (pdyraga) [reply solidity/random-beacon/test/RandomBeacon.GroupCreation.test.ts:1200]: Done in https://github.com/keep-network/keep-core/pull/2734

**Pull request merged**
